### DISCUSSION TITLE
Component throws StreamLostError after some time of dequeuing

### DISCUSSION THREAD

**Discussion by alessandro-gandolfi-ors** [pika/pika#1498] [Q&A]
Hello, I'm having a similar problem similar to the one reported by other users, with a component using Pika that throws StreamLostError. It works for some time, maybe some hours, then it stops completely but I can't understand why.

Here is the stacktrace of the component using Pika:
```
Traceback (most recent call last):
  File "/app/main.py", line 406, in <module>
    main()
  File "/app/main.py", line 386, in main
    channel.basic_ack(
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 2113, in basic_ack
    self._flush_output()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1336, in _flush_output
    self._connection._flush_output(lambda: self.is_closed, *waiters)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 522, in _flush_output
    raise self._closed_result.value.error
pika.exceptions.StreamLostError: Stream connection lost: ConnectionResetError(104, 'Connection reset by peer')
```

Here is the error logged by Rabbit:
```
[error] <0.4070.2> closing AMQP connection <0.4070.2> (10.1.41.121:57236 -> 10.1.41.113:5672):
[error] <0.4070.2> {writer,send_failed,{error,timeout}}
```

This is part of component that was written by another team that decided to dequeue, process and ack the messages in bulk.
The processing could take even less than a minute, **the ack should happen after the processing**.
The initial bulks are processed without any problems for some hours, after a while the error starts to be thrown.
The error is thrown when trying to ack the bulk of messages processed.
I don't think the processing is done in another thread but I tried to make the period between heartbits higher, it doesn't change anything, the error is thrown anyway after less then a minute after receiving the bulk and processing it.
The component and Rabbit are deployed as pods on a K8s cluster.

The component is based on Python 3.9 and Pika 1.1.0.
Even an advice to update the library could be helpful if it could include any type of bugfixes regarding the problem.
At the moment trying to put the processing in another thread could be a bit too much time-spending so other ways are appreciated, using different threads is still acceptable and welcome if there aren't any other "workarounds".

Thanks in advance.

**Comment by lukebakken** [comment]
```
{writer,send_failed,{error,timeout}}
```

This means that RabbitMQ timed out writing to the TCP socket. This is a _very strange_ error.

How large are the messages that are being transmitted from RabbitMQ to this Python/Pika application?

**Reply by alessandro-gandolfi-ors** [reply]
They all should be less than 20kb in average, 30kb max more or less

If it can help, messages are read in chunks of 200, I'm trying now to lower that number and see if there are any changes

**Reply by lukebakken** [reply]
By "chunks of 200", you mean that you are using a prefetch value of 200, correct?

If you'd like me to continue assisting you, you must provide the following:
* RabbitMQ and Erlang version
* _All_ of your code, or a minimal, runnable code sample that demonstrates how you consume messages and do work. Please provide a repository on GitHub with your code. Please note that I _do not have time_ to get your code running.

**Reply by alessandro-gandolfi-ors** [reply]
Yes, the "chunks" are meant to be the prefetch count.

I can say the versions of RabbitMQ (3.11.13) and ErLang (25.3) but having a minimal runnable code could be a little problem, at least for now with the little time that I have to spend for it.

I will try to update the library first, then spend some hours when I can to get a minimal code project working. I will get back in a few days with it. 

Thanks for the support!

**Reply by lukebakken** [reply]
Just so you know, those versions of RabbitMQ and Erlang are both unsupported. I can assist with your code, because the evidence so far points to how you're using Pika.

**Reply by alessandro-gandolfi-ors** [reply]
For now we have found a workaround for the error and it consists in lowering the prefetch count to 25, instead of 200. 50 still threw the error.
Basically the connection just drops during the processing. For now there aren't any logics to establish the connection again.
As for the size of messages, some new data we have received and is used to generate messages is really bigger than usual, reaching even 3mb, but tests ran in local (the setup is sadly not that simple, otherwise I would have already sent the code you requested) with larger messages repeated doesn't seem to be the problem for the connection dropping.
I tried to update Pika and Rabbit to 1.3.2 and 3.13.7 respectively but the problem remains.
